crates/turbo-tasks-macros/src/derive/task_input_macro.rs

```diff
@@ -0,0 +1,267 @@
+                             bounds yet",
+                        )
+                        .emit();
+                }
+            }
+            syn::GenericParam::Lifetime(param) => {
+                param
+                    .span()
+                    .unwrap()
+                    .error("the TaskInput derive macro does not support generic lifetimes")
+                    .emit();
+            }
+            syn::GenericParam::Const(param) => {
+                // NOTE(alexkirsz) Ditto: not supported yet for simplicity's sake.
+                param
+                    .span()
+                    .unwrap()
+                    .error("the TaskInput derive macro does not support const generics yet")
+                    .emit();
+            }
+        }
+    }
+
+    let (try_from_impl, from_impl) = match &derive_input.data {
+        Data::Enum(DataEnum { variants, .. }) => {
+            let mut variants_idents = vec![];
+            let mut variants_fields_len = vec![];
+            let mut variants_fields_destructuring = vec![];
+            let mut variants_try_from_expansion = vec![];
+            let mut variants_from_expansion = vec![];
+
+            for variant in variants {
+                let variant_ident = &variant.ident;
+                let (fields_destructuring, try_from_expansion, from_expansion) = expand_fields(
+                    &variant.ident,
+                    &variant.fields,
+                    &expand_named,
+                    &expand_unnamed,
+                    &expand_unit,
+                );
+                variants_idents.push(variant_ident);
+                variants_fields_len.push(variant.fields.len());
+                variants_fields_destructuring.push(fields_destructuring);
+                variants_try_from_expansion.push(try_from_expansion);
+                variants_from_expansion.push(from_expansion);
+            }
+
+            // This is similar to what Rust does for enums (configurable via the `repr`
+            // attribute). We use the smallest possible integer type that can
+            // represent all the variants. However, for now, this is not
+            // configurable for TaskInput enums.
+            let repr_bits = usize::BITS - variants.len().leading_zeros();
+            let repr = match repr_bits {
+                0..=8 => quote! { u8 },
+                9..=16 => quote! { u16 },
+                17..=32 => quote! { u32 },
+                33..=64 => quote! { u64 },
+                _ => panic!("too many variants"),
+            };
+
+            let variants_discriminants: Vec<_> = (0..variants_idents.len())
+                .map(|i| Literal::usize_unsuffixed(i))
+                .collect();
+
+            (
+                quote! {
+                    match value {
+                        turbo_tasks::TaskInput::List(value) => {
+                            let mut iter = value.iter();
```

Comment:
Nit: make this name more distinct, since it's declaration and usage are split between two different token streams.